File: swift/mflix-movies/Models/SyncStatus.swift
import Foundation

struct SyncStatusInfo: Identifiable, Equatable {
    let id: String
    let isDittoServer: Bool
    let syncSessionStatus: String
    let syncedUpToLocalCommitId: Int?
    let lastUpdateReceivedTime: TimeInterval?
    
    init(from dictionary: [String: Any]) {
        self.id = dictionary["_id"] as? String ?? UUID().uuidString
        self.isDittoServer = dictionary["is_ditto_server"] as? Bool ?? false
        
        if let documents = dictionary["documents"] as? [String: Any] {
            self.syncSessionStatus = documents["sync_session_status"] as? String ?? "Unknown"
            self.syncedUpToLocalCommitId = documents["synced_up_to_local_commit_id"] as? Int
            self.lastUpdateReceivedTime = documents["last_update_received_time"] as? TimeInterval
        } else {
            self.syncSessionStatus = "Unknown"
            self.syncedUpToLocalCommitId = nil
            self.lastUpdateReceivedTime = nil
        }
    }
    
    init?(_ data: Data) {
        do {
            if let jsonObject = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                self.init(from: jsonObject)
            } else {
                return nil
            }
        } catch {
            print("SyncStatusInfo DECODING ERROR:", error.localizedDescription)
            print("📊 Raw data: \(String(data: data, encoding: .utf8) ?? "Unable to convert to string")")
            return nil
        }
    }

    static func == (lhs: SyncStatusInfo, rhs: SyncStatusInfo) -> Bool {
        // Compare the properties that define equality
        return lhs.id == rhs.id &&
        lhs.peerType == rhs.peerType &&
        // Include other properties that define equality
        lhs.syncSessionStatus == rhs.syncSessionStatus &&
        lhs.syncedUpToLocalCommitId == rhs.syncedUpToLocalCommitId
    }

    var formattedLastUpdate: String {
        guard let lastUpdateReceivedTime = lastUpdateReceivedTime else {
            return "Never"
        }
        
        let date = Date(timeIntervalSince1970: lastUpdateReceivedTime / 1000.0)
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .medium
        formatter.doesRelativeDateFormatting = true
        
        return formatter.string(from: date)
    }
    
    var statusColor: String {
        switch syncSessionStatus {
        case "Connected":
            return "green"
        case "Connecting":
            return "orange"
        case "Disconnected":
            return "red"
        default:
            return "gray"
        }
    }
    
    var peerType: String {
        return isDittoServer ? "Cloud Server" : "Peer Device"
    }
}

File: flutter/lib/models/sync_status.dart
class SyncStatus {
  final String id;
  final bool isDittoServer;
  final String syncSessionStatus;
  final int? syncedUpToLocalCommitId;
  final int? lastUpdateReceivedTime;

  SyncStatus({
    required this.id,
    required this.isDittoServer,
    required this.syncSessionStatus,
    this.syncedUpToLocalCommitId,
    this.lastUpdateReceivedTime,
  });

  factory SyncStatus.fromJson(Map<String, dynamic> json) {
    final documents = json['documents'] as Map<String, dynamic>? ?? {};
    
    return SyncStatus(
      id: json['_id'] ?? '',
      isDittoServer: json['is_ditto_server'] ?? false,
      syncSessionStatus: documents['sync_session_status'] ?? '',
      syncedUpToLocalCommitId: (documents['synced_up_to_local_commit_id'] as num?)?.toInt(),
      lastUpdateReceivedTime: (documents['last_update_received_time'] as num?)?.toInt(),
    );
  }

  bool get isConnected => syncSessionStatus == 'Connected';
  
  String get peerType => isDittoServer ? 'Cloud Server' : 'Peer Device';
  
  bool get hasSyncedCommit => syncedUpToLocalCommitId != null;
  
  bool get hasLastUpdateTime => lastUpdateReceivedTime != null;
}